### PR TITLE
Make LstmDataV1 an enum

### PR DESCRIPTION
**Pull request by sffc**
Depends on #3250
Potentially the last part of #1426

This is based on the design of `PropertyCodePointSetV1`.

### CODE REVIEW

File: experimental/segmenter/src/provider.rs [hunk 1]
```diff
@@ -132,7 +132,7 @@ pub struct LstmMatrix<'data> {
 }
 
 #[cfg(feature = "lstm")]
-impl<'data> LstmMatrix<'data> {
+impl<'data> ZeroMatrix32<'data> {
```
Comment (robertbastian) [inline experimental/segmenter/src/provider.rs:135]: nit: `ZeroMatrix32` and `MatrixZero<D>` are confusing names. Maybe we should unify them...

Comment (robertbastian) [reply experimental/segmenter/src/provider.rs:135]: If we are going to unify them then no, otherwise I'd like better names.

Comment (sffc) [reply experimental/segmenter/src/provider.rs:135]: Well your PR touches them, too, and there's a thread over there where you made a case for them to stay separate

Comment (robertbastian) [reply experimental/segmenter/src/provider.rs:135]: I do like having them separate. I could see them having the *same* name, just as `provider::ZeroMatrix` and `math::ZeroMatrix`. That'd be less confusing than trying to figure out what the difference between a `ZeroMatrix` and a `MatrixZero` is.

Comment (sffc) [reply experimental/segmenter/src/provider.rs:135]: I don't see changing the name of the internal (non provider) matrix types as in scope for this PR

Comment (robertbastian) [reply experimental/segmenter/src/provider.rs:135]: Well you're producing a name collision, so I don't see why not. The names have only been there for like 2 days.

Comment (sffc) [reply experimental/segmenter/src/provider.rs:135]: There is

VarZeroVecOwned
Cow::Owned
ZeroVec::Owned
serde::DeserializeOwned

but "Zero" usually goes at the beginning in ZeroVec, ZeroMap, ...

so how about,

- math_helper::Matrix
- math_helper::MatrixOwned
- math_helper::MatrixBorrowed
- math_helper::ZeroMatrixBorrowed
- provider::ZeroMatrix32

Comment (robertbastian) [reply experimental/segmenter/src/provider.rs:135]: sg

**Comment by sffc** [comment]
So much has changed in the last 36 hours that I just made a new PR

https://github.com/unicode-org/icu4x/pull/3264

**Comment by robertbastian** [comment]
My bad

**Pull request closed without merging**